driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -150,13 +170,73 @@ private void buildTableMetadata(KeyspaceMetadata ksm, List<Row> cfRows, Map<Stri
                         cols = Collections.<String, ColumnMetadata.Raw>emptyMap();
                     }
                 }
-                TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+
+                TableMetadata tm = TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+                TableMetadata tmPrevious = ksm.tables.get(tm.getName());
+                if (tmPrevious == null) {
+                    ksm.add(tm);
+                    triggerOnTableAdded(tm);
```

Comment:
There is one caller of this code (refreshSchema via ControlConnection#tryConnect) which can be called in a separate executor than the main cluster executor (either the main thread or in reconnectionExecutor) are we worried about schema listeners blocking those threads?